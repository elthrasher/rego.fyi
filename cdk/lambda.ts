import { Code, Function as LambdaFunction, Runtime } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { BundlingDockerImage, Stack } from '@aws-cdk/core';
import { execSync, ExecSyncOptions } from 'child_process';
import { join } from 'path';

const getAuthorizer = (stack: Stack): LambdaFunction => {
  const execOptions: ExecSyncOptions = { stdio: ['ignore', process.stderr, 'inherit'] };
  const goPath = join(__dirname, '../lambda');
  return new LambdaFunction(stack, 'AuthZFun', {
    code: Code.fromAsset(goPath, {
      bundling: {
        command: ['sh', '-c', 'echo "Docker build not supported. Please install go."'],
        image: BundlingDockerImage.fromRegistry('alpine'),
        local: {
          tryBundle(outputDir: string) {
            try {
              execSync('go version', execOptions);
            } catch {
              return false;
            }
            execSync(`GOARCH=amd64 GOOS=linux go build -ldflags="-s -w" -o ${join(outputDir, 'main')}`, {
              ...execOptions,
              cwd: goPath,
            });
            return true;
          },
        },
      },
    }),
    handler: 'main',
    runtime: Runtime.GO_1_X,
  });
};

const getLambdalith = (stack: Stack): LambdaFunction =>
  new NodejsFunction(stack, 'LambdalithFn', {
    entry: `${__dirname}/../lambda/lambdalith.ts`,
    runtime: Runtime.NODEJS_14_X,
  });

export const getFns = (stack: Stack): { [fnName: string]: LambdaFunction } => ({
  authorizer: getAuthorizer(stack),
  lambdalith: getLambdalith(stack),
});
