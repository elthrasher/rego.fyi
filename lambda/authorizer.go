package main

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"strings"

	"fmt"
	"log"
	"time"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/open-policy-agent/opa/ast"
	"github.com/open-policy-agent/opa/rego"
	"github.com/open-policy-agent/opa/storage"
	"github.com/open-policy-agent/opa/storage/inmem"
	"github.com/open-policy-agent/opa/util"
)

var module string
var store storage.Store
var compiler *ast.Compiler
var ctx context.Context

func main() {
	log.Println("cold start")

	// just compile OPA policies once per container

	//os.Exit(0)
	lambda.Start(Handler)
}

func track(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}

func compileOpaPolicy(module string, data string) {
	defer track(time.Now(), "compileOpaPolicy()")

	ctx = context.Background()

	store = inmem.NewFromReader(bytes.NewBufferString(data))

	// Parse the module. The first argument is used as an identifier in error messages.
	parsed, err := ast.ParseModule("entitlements.rego", module)
	if err != nil {
		// Handle error.
	}

	// Create a new compiler and compile the module. The keys are used as
	// identifiers in error messages.
	compiler = ast.NewCompiler()
	compiler.Compile(map[string]*ast.Module{
		"entitlements.rego": parsed,
	})

	if compiler.Failed() {
		// Handle error. Compilation errors are stored on the compiler.
		panic(compiler.Errors)
	}
}

func checkOpaPolicy(method string, permissions []string, resourcePath string, subscriptions []string) (result bool) {
	defer track(time.Now(), "checkOpaPolicy()")

	// Create a new query that uses the compiled policy from above.
	reg := rego.New(
		rego.Query("data.entitlements.allow"),
		rego.Store(store),
		rego.Compiler(compiler),
		rego.Input(
			map[string]interface{}{
				"method":        method,
				"permissions":   permissions,
				"resource":      resourcePath,
				"subscriptions": subscriptions,
			},
		),
	)

	// Run evaluation.
	rs, err := reg.Eval(ctx)

	if err != nil {
		log.Println("err:", err)
	}

	if len(rs) != 1 {
		return false
	}

	result = util.Compare(rs[0].Expressions[0].Value, true) == 0

	return result

}

// Handler is the stuff
func Handler(event events.APIGatewayCustomAuthorizerRequestTypeRequest) (events.APIGatewayCustomAuthorizerResponse, error) {
	defer track(time.Now(), "Handler()")
	log.Println("handle event")

	/**
	check Authorization header first - just using 'allow' and 'deny' value to simulate for now
	JWT validation can replace this to check signing/exp, and then auth caching can be enabled in apigw
	*/
	rawToken := strings.ReplaceAll(event.Headers["Authorization"], "Bearer ", "")

	decodedStr, err := base64.StdEncoding.DecodeString(rawToken)
	if err != nil {
		fmt.Printf("failed to parse payload: %s\n", err)
	}

	var dat map[string]string

	if err := json.Unmarshal(decodedStr, &dat); err != nil {
		fmt.Printf("Could not deserialize json: %s\n", err)
	}

	var token map[string][]string

	json.Unmarshal([]byte(dat["token"]), &token)

	compileOpaPolicy(dat["rego"], dat["data"])

	resourcePath := event.Path
	if checkOpaPolicy(event.HTTPMethod, token["permissions"], resourcePath, token["subscriptions"]) {
		log.Println("OPA policy check ok - request allowed")
		return generateIAMPolicy("user", "Allow", event.MethodArn), nil
	}
	log.Println("failed OPA policy check")
	return generateIAMPolicy("user", "Deny", event.MethodArn), nil

}

/**
Generate IAM policy document
*/
func generateIAMPolicy(principalID string, effect string, resource string) events.APIGatewayCustomAuthorizerResponse {
	defer track(time.Now(), "generateIAMPolicy()")
	authResponse := events.APIGatewayCustomAuthorizerResponse{PrincipalID: principalID}

	if effect != "" && resource != "" {
		authResponse.PolicyDocument = events.APIGatewayCustomAuthorizerPolicy{
			Version: "2012-10-17",
			Statement: []events.IAMPolicyStatement{
				{
					Action:   []string{"execute-api:Invoke"},
					Effect:   effect,
					Resource: []string{resource},
				},
			},
		}
	}

	authResponse.Context = map[string]interface{}{}
	return authResponse
}
