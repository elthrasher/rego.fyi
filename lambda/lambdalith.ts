import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';

export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
  const { httpMethod, path } = event;
  return {
    body: JSON.stringify({ httpMethod, path }),
    headers: { 'Access-Control-Allow-Credentials': true, 'Access-Control-Allow-Origin': 'https://rego.fyi' },
    statusCode: 200,
  };
};
