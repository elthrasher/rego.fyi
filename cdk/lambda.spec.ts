import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { getFns } from './lambda';

describe('lambda builds', () => {
  test('get functions', () => {
    const app = new App();
    const stack = new Stack(app);
    getFns(stack);
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
  });
});
