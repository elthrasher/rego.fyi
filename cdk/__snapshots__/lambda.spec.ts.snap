// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lambda builds get functions 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "AuthZFun2B67BA45": Object {
      "DependsOn": Array [
        "AuthZFunServiceRoleFC1807CF",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Handler": "main",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AuthZFunServiceRoleFC1807CF",
            "Arn",
          ],
        },
        "Runtime": "provided.al2",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AuthZFunServiceRoleFC1807CF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdalithFnE6677028": Object {
      "DependsOn": Array [
        "LambdalithFnServiceRoleEAD0A948",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdalithFnServiceRoleEAD0A948",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdalithFnServiceRoleEAD0A948": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
